## 1. a) Give the difference between type casting and automatic type conversion. Also, give a suitable Python code to illustrate both.

| Feature | Type Casting | Automatic Type Conversion |
|-|-|-|
| **Definition** | Explicit conversion by the programmer. | Implicit conversion by Python (PVM). |
| **Control** | Requires explicit instructions (e.g., `int()`, `float()`). | Happens automatically during operations. |
| **Example**  | `int("123")` converts string to integer.  | Adding `5 + 3.2` converts `5` to float. |

---

## 1. b) Describe about syntax error, runtime error and logical error.

| Aspect                | **Syntax Errors**                          | **Runtime Errors**                         | **Semantic (Logical) Errors**              |
|-----------------------|--------------------------------------------|-------------------------------------------|--------------------------------------------|
| **Definition**        | Errors in the code structure or syntax that prevent execution. | Errors that occur during program execution (Exceptions). | Errors in logic that lead to incorrect results despite successful execution. |
| **When Detected**     | Before execution (During the compilation or interpretation phase). | During excution.              | After execution (During testing phase). |
| **Cause**             | Violations of language rules. | Invalid operations (e.g., division by zero, accessing undefined variables). | Flaws in program logic/Algorithm leading to unexpected ouputs. |
| **Examples**          | `if x = 5:` (should be `if x == 5:`)       | `print(5 / 0)` (division by zero)          | Calculating the area of a rectangle as `length + width` instead of `length * width`. |
| **Resolution**        | Fix the syntax to adhere to language rules. | Exception handling. | Correct the logic or algorithm.            |


---

## 1. c) Describe various features of Python.

1. **Easy to learn and use** – Simple and intuitive syntax.  
2. **Interpreted** – Executes code line-by-line (Technically compiled then interpreted).  
3. **Dynamically typed** – No need for explicit variable type declaration.  
4. **Platform independent** – Works across multiple operating systems.  
5. **Extensive libraries** – Offers built-in and third-party modules.  
6. **Open source** – Free to use, modify, and distribute.  
7. **Object-oriented** – Supports object-oriented programming.  
8. **High-level language** – Abstracts low-level details.  
9. **Multi-paradigm support** – Procedural, object-oriented, and functional programming.  
10. **Automatic memory management** – Handles garbage collection automatically.  

## 1. d) Write an iterative program (non-recursive) that asks the user to enter an integer and computes the factorial Of that integer, usually written n! in mathematics.

In [2]:
def factorial(n:int)->int:
    assert n>=0, "n should be a positive integers."
    f = 1
    for i in range(1, n+1):
        f *= i
    return f

In [3]:
factorial(5)

120

## 2. a) How multiple inheritance is different from multi-level inheritance? Explain with the help of example.


| **Aspect**                | **Multiple Inheritance**                                                                 | **Multi-Level Inheritance**                                                      |
|---------------------------|-----------------------------------------------------------------------------------------|----------------------------------------------------------------------------------|
| **Definition**            | A class inherits from multiple parent classes.                                         | A class inherits from a child class, which in turn inherits from another parent class. |
| **Hierarchy**             | Involves multiple base classes directly inherited by a single class.                   | Involves a chain of inheritance (a class inherits from a child class that inherits from another parent). |
| **Example**               | `class A, B: class C(A, B)`: Class `C` inherits from both `A` and `B`.               | `class A: class B(A): class C(B)`: Class `C` inherits from `B`, which inherits from `A`. |
| **Complexity**            | Can lead to ambiguity if not handled properly (diamond problem).                     | Simpler chain structure, representing a step-by-step inheritance path.           |

In [4]:
# Multiple Inheritence
class A:
    name = "A"
    
class B:
    name = "B"

class C(A,B):
    name = "C"

In [5]:
C.__mro__

(__main__.C, __main__.A, __main__.B, object)

In [6]:
# Multi-level Inheritence
class A:
    name = "A"

class B(A):
    name = "B"

class C(B):
    name = "C"

In [7]:
C.__mro__

(__main__.C, __main__.B, __main__.A, object)

> Python resolves multiple inheritances using the C3 linearization algorithm to determine the Method Resolution Order (MRO). This algorithm ensures that the inheritance order is consistent and that each class appears before its parents.

## 2. b) Explain the main features of an object oriented programming language.

1. **Encapsulation**: Groups data and methods into objects to limit direct access.
2. **Abstraction**: Hides complex implementation details and exposes only essential features.
3. **Inheritance**: Enables new classes to derive attributes and methods from existing ones.
4. **Polymorphism**: Allows methods to have different implementations based on the context.
5. **Modularity**: Promotes reusable, organized, and manageable code.

## 2. c) What advantages do NumPy arrays offer over (nested) Python lists?

| **Feature**               | **NumPy Array**                                      | **Nested Python List**                          |
|----------------------------|-----------------------------------------------------|------------------------------------------------|
| **Performance**            | Faster due to C implementation and vectorization    | Slower as it relies on Python loops            |
| **Memory Efficiency**      | More memory-efficient (contiguous memory storage)   | Less efficient due to overhead of references   |
| **Broadcasting**           | Supports broadcasting for operations on different shapes | Not supported                                  |
| **Convenience**            | Rich set of built-in functions (math, stats, etc.)  | Requires custom implementation for operations  |
| **Data Type Consistency**  | Homogeneous (fixed type)                            | Heterogeneous (mixed types allowed)            |
| **Indexing & Slicing**     | Advanced indexing, slicing, and boolean masking     | Limited to basic slicing and indexing          |

## 2. d) Write a menu driven python program to perform various list operations, such as:
i. Append an element\
ii. Insert an element\
iii. Append a list to the given list\
iv. Modify an existing element\
v. Delete an existing element from its position\
vi. Delete an existing element with a given value\
vii. Sort the list in ascending order\
viii. Sort the list in descending order\
(Hint: Say your list has 5 elements: `[24, 3, 86, 15, 7]`)

In [16]:
l = [24, 3, 86, 15, 7]

while True:
    c = int(input("""1. Append an element\n
                  2. Insert an element\n
                  3. Append a list to the given list\n
                  4. Modify an existing element\n
                  5. Delete an existing element from its position\n
                  6. Delete an existing element with a given value\n
                  7. Sort the list in ascending order\n
                  8. Sort the list in descending order\n
                  9. Exit\n
                  ENTER YOUR CHOICE (1-9): """))
    if c == 9:
        break
    elif c == 1:
        n = int(input("Enter the element to be appended: "))
        l.append(n)
    elif c == 2:
        n = int(input("Enter the element to be inserted: "))
        index = int(input("Enter the index to insert the element at: "))
        l.insert(index, n)
    elif c == 3:
        l1 = input("Enter a comma separated list (e.g for [1, 2] enter 1, 2): ")
        l.append([int(x) for x in l1.split(',')])
    elif c == 4:
        index = int(input("Enter the index of element to be modified: "))
        value = int(input("Enter the new value for the element: "))
        l[index] = value        
    elif c == 5:
        index = int(input("Enter the index of element to be deleted: "))
        l.pop(index)
    elif c == 6:
        value = int(input("Enter the value of the element to be deleted: "))
        l.remove(value)
    elif c == 7:
        l.sort()
    elif c == 8:
        l.sort(reverse=True)
    else:
        print("Error: Choose correct option (1-9).")
    print(l)

[24, 3, 86, 15, 7, 10]


## 3. a) Looking at the below code, write down the final values of A0, Al, A2 and A3.


In [ ]:
A0 = dict(zip(('a', 'b', 'c', 'd', 'e'), (1, 2, 3, 4, 5)))
A1 = range(10)
A2 = {i:i*i for i in A1}
A3 = [[i, i*i] for i in A1]

print(A0) # {'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5}
print(A1) # range(0, 10) <- Range Object
print(A2) # {0: 0, 1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81}
print(A3) # [[0, 0], [1, 1], [2, 4], [3, 9], [4, 16], [5, 25], [6, 36], [7, 49], [8, 64], [9, 81]]

{'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5}
range(0, 10)
{0: 0, 1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81}
[[0, 0], [1, 1], [2, 4], [3, 9], [4, 16], [5, 25], [6, 36], [7, 49], [8, 64], [9, 81]]


## 3. b) Lists and tuples are similar in many ways with one major difference; that lists are mutable and tuples are not. Elaborate this and give an example of an operation that lists support, but tuples do not. Further, discuss, which data structure should be utilized under what circumstances.

- The line "lists are mutable and tuples are not" implies that the elements of a list can be modified where as the elements of a tuple cannot be modified at the same memory location after creation. For instance,
    ```python
    l = [1,2,3]
    l[0] = 5

    t = [1,2,3]
    t[0] = 5
    ```
    Here, l[0] = 5 will update the list l's first element to 5 but, t[0] = 5 will throw error as tuple does not support item assignment.

- Example of operation supported by lists, but not tuple: `append`.
    ```python
    l = [1,2,3]
    l.append(5) # l becomes [1, 2, 3, 4, 5]

    t = [1,2,3]
    t.append(5) # Throws error: Tuples do not support append
    ```
- We should use tuples when we need our data to remain unchanged for example storing coordinates of a geographical location (longitude, attitude), but if we want to change our data in future we should use lists for instance making list of students of a class (expecting new admissions and TCs).


## 3. c) (ii)

In [54]:
import re

regex = r"([a-zA-Z]+) (\d+)"
match = re.search(regex, "I was born on June 24")
if match != None:
    print("Match at index %s, %s" % (match.start(), match.end()) )
    print ("Full match: %s" % (match.group(0)) )
    print ("Month: %s" % (match.group( 1)) )
    print("Day: %s" % (match.group(2)))
else:
    print("The regex pattern does not match.")

Match at index 14, 21
Full match: June 24
Month: June
Day: 24


## 3. d) What does *args mean? Why is it used?

`*args` is a special syntax in python which is used to pass a variable number of positional arguments to a function. Here the '*' is mandatory to write and `args` can be any other vaid name.

**Why is it Used?:**

1. **Flexibility**:
   - It enables a function to accept an arbitrary number of arguments, making it more flexible.
   - Example: A function to sum any number of numbers.

2. **Readability**:
   - It provides a clean and simple way to handle variable arguments instead of using workarounds like a list or manual unpacking.

3. **Reusability**:
   - It avoids the need to overload functions for different numbers of arguments.

**Example:**

```python
def add_numbers(*args):
    return sum(args)

print(add_numbers(1, 2, 3))  # Output: 6
print(add_numbers(10, 20, 30, 40))  # Output: 100
```

- The function can handle any number of inputs without requiring changes to its definition.

## 4 a) Create a class 'employee' for a company that should have the properties like the name of an employee, salary of an employee, employee ID and the designation. In the class, whenever an object of an employee is created, employee ID should be automatically formed. Further the class should have two functions — One function should show the details of the employees and other function should return the total number of employees. With respect to Object Oriented Programming, is there any special type of variable highlighted in the above code. Explain how the variable is different from Instance variable. 

In [60]:
class Employee:
    total_employees = 0

    def __init__(self, name, salary, designation):
        self.name = name 
        self.salary = salary 
        self.designation = designation
        Employee.total_employees += 1 
        # For emp_id we have EMP + first 3 char of designation followed by count of employee
        self.emp_id = f"EMP{designation[:3].upper()}{Employee.total_employees:03}"

    def show_details(self):
        print(f"Employee ID: {self.emp_id}")
        print(f"Name: {self.name}")
        print(f"Salary: {self.salary}")
        print(f"Designation: {self.designation}")

    @classmethod
    def get_total_employees(cls):
        return cls.total_employees


employee1 = Employee("Raj", 90000, "Developer")
employee2 = Employee("Amit", 60000, "Manager")

employee1.show_details()
print()
employee2.show_details()
print("\nTotal Employees:", Employee.get_total_employees())

Employee ID: EMPDEV001
Name: Raj
Salary: 90000
Designation: Developer

Employee ID: EMPMAN002
Name: Amit
Salary: 60000
Designation: Manager

Total Employees: 2


**Special Type of Variable highlighted in the above code: Class Variable**
- A class variable is a variable that is declared inside the class but outside any instance method. 
- It is shared by all instances of the class.
**Difference Between Class and Instance Variables:**

| Feature                   | Instance Variable                           | Class Variable                            |
|---------------------------|---------------------------------------------|-------------------------------------------|
| **Scope**                 | Specific to an object.                     | Shared across all objects of the class.   |
| **Storage**               | Stored in the instance (object) namespace. | Stored in the class namespace.            |
| **Access**                | Accessed with `self.<variable>`.           | Accessed with `ClassName.<variable>` or `self.<variable>`. |
| **Use Case**              | Used to store data unique to each object.  | Used for properties shared by all objects (e.g., counters). |
| **Example**               | `self.name`, `self.salary`                 | `Employee.total_employees`                |

---

## 4. b) What is the purpose of using function in python programming? Give the syntax of defining a function in python.

**Purpose:**
1. **Code Reusability**: Avoid repetition by reusing code in multiple places.
2. **Modularity**: Break a program into smaller, manageable parts.
3. **Maintainability**: Simplify debugging and updating code.
4. **Readability**: Improve understanding of the program structure.
5. **Scalability**: Easily handle growing complexity by dividing logic into functions.

**Syntax of Defining a Function in Python:**

```python
def function_name(parameters):
    """Optional docstring"""
    # Function body
    return value  # Optional
```

## 4. c) What are Lambda functions? Describe.

Lambda functions are **anonymous, single-line functions** defined using the `lambda` keyword in Python. They are used to write quick, small functions without formally defining them using `def`.


**Syntax:**
```python
lambda arguments: expression
```
- **`arguments`**: Inputs to the function.
- **`expression`**: Single expression evaluated and returned.


**Example:**
```python
square = lambda x: x**2
print(square(5))  # Output: 25
```

## 4. d) What do you understand by local and global scope of variables? How can you access a global variable inside the function, if function has a variable with same name?

Variabls can be of local or global scope. If they are defined at the top level of a script (outside any function) they have **global scope**, whereas if they are defined in the body of any funciton then they have **local scope**.

If the funciton has a variable with same name as that of a global variable then the global variable has to be accessed using the `global` keyword.


In [62]:
a = 1
print(a)
def fun():
    global a
    a = 5
    print(a)
fun()
print(a)

1
5
5


---

## 5. a) Write a program to calculate in how many days a work will be completed by three persons A, B and C together. A, B, C take x days, y days and z days respectively to do the job alone. The formula to calculate the number of days if they work together is xyz/(xy + yz + xz) days where x, y, and z are given as input to the program.

In [65]:
x = int(input("Enter number of days taken by A: ")) # 10
y = int(input("Enter number of days taken by B: ")) # 5
z = int(input("Enter number of days taken by C: ")) # 10

d = (x*y*z)/(x*y+y*z+x*z)

print(f"Number of days required for A, B and C to complete the work together is {d:.2f}")

Number of days required for A, B and C to complete the work together is 2.50


## 5. b) Compare recursive and iterative techniques for problem solving.

| **Aspect**        | **Recursion**                               | **Iteration**                          |
|--------------------|-----------------------------------------------|----------------------------------------|
| **Definition**     | Function calls itself to solve subproblems.  | Uses loops (for/while) to repeat logic. |
| **Memory Usage**   | Uses more memory due to the call stack.      | Uses less memory by reusing the same memory space. |
| **Performance**    | Slower due to repeated calls and stack operations. | Generally faster and efficient. |
| **Debugging**      | Harder to debug.                            | Easier to debug.                      |
| **Complexity**     | Can be simpler for hierarchical/recursive problems. | Simpler for straightforward problems. |
| **Example (Factorial)** | `def factorial(n):`<br>&nbsp;&nbsp;&nbsp;`if n == 0:`<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`return 1`<br>&nbsp;&nbsp;&nbsp;`else:`<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`return n * factorial(n - 1)`<br>`print(factorial(5))  # Output: 120` | `def factorial(n):`<br>&nbsp;&nbsp;&nbsp;`result = 1`<br>&nbsp;&nbsp;&nbsp;`for i in range(1, n + 1):`<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`result *= i`<br>&nbsp;&nbsp;&nbsp;`return result`<br>`print(factorial(5))  # Output: 120` |